python/demo/demo_pml.py

```diff
@@ -0,0 +1,225 @@
+import sys
+
+try:
+    import gmsh
+except ModuleNotFoundError:
+    print("This demo requires gmsh to be installed")
+    sys.exit(0)
+import numpy as np
+
+try:
+    import pyvista
+    have_pyvista = True
+except ModuleNotFoundError:
+    print("pyvista and pyvistaqt are required to visualise the solution")
+    have_pyvista = False
+from data.mesh_wire_pml import generate_mesh_wire
+
+import ufl
+from data.utils import calculate_analytical_efficiencies
+from dolfinx import fem, plot
+from dolfinx.io import VTXWriter
+from dolfinx.io.gmshio import model_to_mesh
+from ufl import FacetNormal, as_matrix, as_vector, cross, curl, inner, lhs, rhs, sqrt, algebra, \
+    transpose, det, inv, grad
+
+from mpi4py import MPI
+from petsc4py import PETSc
+
+if not np.issubdtype(PETSc.ScalarType, np.complexfloating):
+    print("Demo should only be executed with DOLFINx complex mode")
+    exit(0)
+
+class BackgroundElectricField:
+
+    def __init__(self, theta, n_b, k0):
+        self.theta = theta
+        self.k0 = k0
+        self.n_b = n_b
+
+    def eval(self, x):
+
+        kx = self.n_b * self.k0 * np.cos(self.theta)
+        ky = self.n_b * self.k0 * np.sin(self.theta)
+        phi = kx * x[0] + ky * x[1]
+
+        ax = np.sin(self.theta)
+        ay = np.cos(self.theta)
+
+        return (-ax * np.exp(1j * phi), ay * np.exp(1j * phi))
+
+def curl_2d(a):
+    """Returns the curl of two 2D vectors as a 3D vector"""
+    ay_x = a[1].dx(0)
+    ax_y = a[0].dx(1)
+    return as_vector((0, 0, ay_x - ax_y))
+
+def pml_coordinates(x, alpha, k0, l_dom, l_pml):
+
+    # Define boolean functions for the PML regions
+    inside_pml_x = (ufl.sign(ufl.sign(x[0])*x[0] - l_dom/2) + 1)/2
+    inside_pml_y = (ufl.sign(ufl.sign(x[1])*x[1] - l_dom/2) + 1)/2
+
+    # Define the coordinate transformation for PML regions
+    x_pml = x[0] + 1j*alpha/k0*x[0]*(ufl.sign(x[0])*x[0] - l_dom/2)/(l_pml/2 - l_dom/2)**2*inside_pml_x
+    y_pml = x[1] + 1j*alpha/k0*x[1]*(ufl.sign(x[1])*x[1] - l_dom/2)/(l_pml/2 - l_dom/2)**2*inside_pml_y
```

Comment:
```suggestion
    x_pml = x[0] + 1j*alpha/k0*x[0]*(ufl.sign(x[0])*x[0] - l_dom/2)/(l_pml/2 - l_dom/2)**2*inside_pml_x
    y_pml = x[1] + 1j*alpha/k0*x[1]*(ufl.sign(x[1])*x[1] - l_dom/2)/(l_pml/2 - l_dom/2)**2*inside_pml_y
```
Instead of repeating `x_pml` and `y_pml`, you can use `ufl.Index()`, as illustrated in the following minimal example:
```python
import dolfinx
from mpi4py import MPI
import ufl

mesh = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, 10, 10)
V = dolfinx.fem.FunctionSpace(mesh, ("CG", 1))
u = dolfinx.fem.Function(V)
u.interpolate(lambda x: x[0])
x = ufl.SpatialCoordinate(mesh)


def F(u, x):
    qx = 2 * u * x[0]
    qy = 2 * u * x[1]
    return ufl.as_vector((qx, qy))


form1 = dolfinx.fem.form(ufl.inner(F(u, x), F(u, x))*ufl.dx)
print(dolfinx.fem.assemble_scalar(form1))


def G(u, x):
    i = ufl.Index()
    qi = 2 * u * x[i]
    return ufl.as_tensor(qi, i)


Gi = G(u, x)
form2 = dolfinx.fem.form(ufl.inner(Gi, Gi)*ufl.dx)
print(dolfinx.fem.assemble_scalar(form2))
```
returning
```bash
1.244444444444446
1.244444444444446
```